### ISSUE TITLE
ill-typed generated projectors causes F* to reject valid and useful inductive type (cexists_type_to_type)

### ISSUE THREAD

**Issue by catalin-hritcu** [FStarLang/FStar#350]
Because of the lack of kind polymorphism we need to manually write constructive existentials for each kind, but at the moment this only works for kind `Type` (see `cexists_type` in `constr.fst`). The following generalization to kind `Type->Type`

```
type cexists_type_to_type : ((Type->Type) -> Type) -> Type =
  | ExTypeToTypeIntro : #p:((Type->Type) -> Type) -> t:(Type -> Type) ->
                         h:(p t) -> cexists_type_to_type p
```

is rejected with an obscure error message:

```
inj.fst(28,4-28,21): Subtyping check failed; expected type (p projectee (t projectee)); got type (_2_12 _2_13)
```

with `--print_implicits` this error message looks as follows:

```
inj.fst(28,4-28,21): Subtyping check failed; expected type (p #(fun (Type -> Type) -> (_0 _)) projectee (t #(fun (Type -> Type) -> (_0 _)) projectee)); got type (_2_12 _2_13)
```

**Comment by catalin-hritcu** [comment]
The current workaround (in `paradoxes/inj.fst`) is to assume this type, it's constructor and it's induction principle, and F* accepts all this just fine.
```
assume type cexists_type_to_type : ((Type->Type) -> Type) -> Type
assume val exTypeToTypeIntro : #p:((Type->Type) -> Type) -> t:(Type -> Type) ->
                                h:(p t) -> Tot (cexists_type_to_type p)
assume val exInd : #p:((Type->Type) -> Type) -> p0:Type ->
             (x:(Type->Type) -> p x -> Tot p0) -> cexists_type_to_type p -> Tot p0
```

**Comment by nikswamy** [comment]
Removed the assumes. 

But added the `--__temp_no_proj` flag to workaround a known bug in the projector code-generation. What you were observing was a failure to type-check the generated code for the projector ... the definition of `cexists_type_to_type` itself is well-formed.

**Comment by kyoDralliam** [comment]
Fixed in 971ea45 with the elaboration of projectors after typechecking.
